File: fixtures/scale/java/large/src/cryptobench/ciphers/xorshift/XorShiftCipher.java
package cryptobench.ciphers.xorshift;

import cryptobench.core.Cipher;
import cryptobench.util.Bytes;
import cryptobench.util.Hex;

/** Keystream from a xorshift64 generator seeded by the key. */
public class XorShiftCipher implements Cipher {
    private final XorShiftKey key;

    public XorShiftCipher(XorShiftKey key) {
        this.key = key;
    }

    @Override
    public String name() {
        return "xorshift";
    }

    @Override
    public String encrypt(String plaintext) {
        return Hex.encode(mask(Bytes.of(plaintext)));
    }

    @Override
    public String decrypt(String ciphertext) {
        return Bytes.toText(mask(Hex.decode(ciphertext)));
    }

    private byte[] mask(byte[] data) {
        long state = key.getSeed();
        byte[] out = new byte[data.length];
        for (int i = 0; i < data.length; i++) {
            state ^= state << 13; state ^= state >>> 7; state ^= state << 17;
            out[i] = (byte) (data[i] ^ (state >>> 16));
        }
        return out;
    }
}

File: fixtures/scale/dart/large/lib/ciphers/xorshift/xorshift_cipher.dart
import 'dart:typed_data';

import '../../core/cipher.dart';
import '../../util/bytes.dart';
import '../../util/hex.dart';
import 'xorshift_key.dart';

/// Keystream from a xorshift32 generator seeded by the key.
class XorShiftCipher implements Cipher {
  final XorShiftKey key;

  XorShiftCipher(this.key);

  @override
  String name() => 'xorshift';

  @override
  String encrypt(String plaintext) => hexEncode(_mask(bytesOf(plaintext)));

  @override
  String decrypt(String ciphertext) => bytesToText(_mask(hexDecode(ciphertext)));

  Uint8List _mask(Uint8List data) {
    var state = key.seed;
    final out = Uint8List(data.length);
    for (var i = 0; i < data.length; i++) {
      state ^= (state << 13) & 0xFFFFFFFF; state ^= (state >> 7); state ^= (state << 17) & 0xFFFFFFFF; state &= 0xFFFFFFFF;
      out[i] = data[i] ^ ((state >> 16) & 0xFF);
    }
    return out;
  }
}
